genie-agent/src/main/java/com/netflix/genie/agent/execution/statemachine/actions/MonitorJobAction.java

```diff
@@ -62,10 +62,30 @@ protected Events executeStateAction(final ExecutionContext executionContext) {
             throw new RuntimeException("Interrupted while waiting for job completion", e);
         }
 
-        log.info("Job process completed with exit code: {}", exitCode);
+        try {
+            // Evil-but-necessary little hack.
+            // The agent and the child job process receive SIGINT at the same time.
+            // If the child terminates quickly, the code below will execute before the signal handler has a chance to
+            // set the job as killed, and the final status would be (incorrectly) reported as failed (due to non-zero
+            // exit code).
+            // So give the handler a chance to mark the context before attempting to read it.
```

Comment:
Instead, why cant this code wait on a condition which will be fulfilled only when the job is marked as killed.

Reply:
There's no guarantee that will happen, or happen in time.
So we could replace the `sleep(1000)` with `latch.wait(1000)` or similar. But in the end, the two are the same thing ("wait a little and hope the other thread does its thing").
Plus we have to share the latch between here and there.

Reply:
`Thread.sleep` just seems arbitrary. With a rightly named latch and a timeout it becomes more obvious what one is doing. But you can argue that the comment you gave provides that context. I would make a call by thinking that if this is the only condition then having a sleep can be made to work. If there are more conditions in the future then having a latch would be more appropriate. For now its fine.

Reply:
Let's say we have a latch and we count it down in the signal handler.
And here we have `latch.wait(1000)`.

Case 1: job is killed: this code waits for the latch.
If it gets decremented in time, great. If not we timeout and miss the fact this was a kill.
But `sleep` provides the same guarantees.

Case 2: job is not killed.
This code waits for the latch, then times out (nobody is going to decrement it!) and moves on.
Once again, this is the same as what 'sleep' would have done.

The only benefit of a latch is waiting slightly less in case of kill (i.e. wake up and move on before timing out).